src/util.c

```diff
@@ -487,55 +490,93 @@ systemd_escape(char* string)
         // LCOV_EXCL_STOP
     }
-
-    if (!np_state->netdefs) {
+    if (!netplan_util_create_yaml_patch(yaml_path, "NULL", patch_fd, &error)) {
         // LCOV_EXCL_START
-        g_fprintf(stderr, "netplan_delete_connection: %s\n", error->message);
+        g_fprintf(stderr, "netplan_delete_connection: Cannot create YAML patch: %s\n", error->message);
         ret = FALSE;
         goto cleanup;
         // LCOV_EXCL_STOP
     }
 
-    /* find filename for specified netdef ID */
-    nd = g_hash_table_lookup(np_state->netdefs, id);
-    if (!nd) {
-        g_warning("netplan_delete_connection: Cannot delete %s, does not exist.", id);
+    /* Create a new parser & state to hold our output YAML, ignoring the to be
+     * deleted Netdef from the patch */
+    output_parser = netplan_parser_new();
+    output_state = netplan_state_new();
+
+    lseek(patch_fd, 0, SEEK_SET);
+    if (   !netplan_parser_load_nullable_fields(output_parser, patch_fd, &error)
+        || !netplan_parser_load_yaml_hierarchy(output_parser, rootdir, &error)) {
+        // LCOV_EXCL_START
+        g_fprintf(stderr, "netplan_delete_connection: Cannot load output state: %s\n", error->message);
         ret = FALSE;
         goto cleanup;
+        // LCOV_EXCL_STOP
     }
 
-    del = g_strdup_printf("network.%s.%s=NULL", netplan_def_type_name(nd->type), id);
+    lseek(patch_fd, 0, SEEK_SET);
+    if (!netplan_parser_load_yaml_from_fd(output_parser, patch_fd, &error)) {
+        // LCOV_EXCL_START
+        g_fprintf(stderr, "netplan_delete_connection: Cannot parse YAML patch: %s\n", error->message);
+        ret = FALSE;
+        goto cleanup;
+        // LCOV_EXCL_STOP
+    }
 
-    /* TODO: refactor logic to actually be inside the library instead of spawning another process */
-    const gchar *argv[] = { SBINDIR "/" "netplan", "set", del, NULL, NULL, NULL };
-    if (rootdir) {
-        argv[3] = "--root-dir";
-        argv[4] = rootdir;
+    /* We're only deleting some data, so FALLBACK_FILENAME should never be created */
+    if (   !netplan_state_import_parser_results(output_state, output_parser, &error)
+        || !netplan_state_update_yaml_hierarchy(output_state, FALLBACK_FILENAME, rootdir, &error)) {
+        // LCOV_EXCL_START
+        g_fprintf(stderr, "netplan_delete_connection: Cannot write output state: %s\n", error->message);
+        ret = FALSE;
+        goto cleanup;
+        // LCOV_EXCL_STOP
     }
-    if (getenv("TEST_NETPLAN_CMD") != 0)
-       argv[0] = getenv("TEST_NETPLAN_CMD");
-    ret = g_spawn_sync(NULL, (gchar**)argv, NULL, 0, NULL, NULL, NULL, NULL, NULL, NULL);
 
 cleanup:
-    if (npp) netplan_parser_clear(&npp);
-    if (np_state) netplan_state_clear(&np_state);
+    if (input_parser) netplan_parser_clear(&input_parser);
+    if (input_state) netplan_state_clear(&input_state);
+    if (output_parser) netplan_parser_clear(&output_parser);
+    if (output_state) netplan_state_clear(&output_state);
+    if (patch_fd > 0) close(patch_fd);
```

Comment:
It should be `patch_fd >= 0`. If the file descriptor 0 doesn't exist when memfd_create() is called, 0 will be used and the function would leak the descriptor (and memory).